### PR TITLE
Atari New server

### PR DESCRIPTION
**Pull request by fchirica**
Manage data workflow:

Every `manage_data_interval` do the following in order:

1. Iterate over all tree_ids in local store, and if we're not subscribed to them, try to subscribe (assuming the corresponding singleton is on chain). Subscriptions will contain our data as well as other tree_ids we're interested in. 

2. Iterate over all subscriptions, and if the local store generation is behind the on chain generation, try to "sync":

Let A the local generation and B the on-chain target generation. Take the first server in order of preference and download delta file for generation A+1. Then, try to insert into the local store the root corresponding to generation A+1. If successful, generate the full tree file for generation A+1, and then try A+2, A+3, ... until we get to target generation B. If failure at any step, move to the next server in order of preference, starting from the last successfully validated generation using the current server. At the end of the "sync", we'll have all the delta files (as we've downloaded them from other servers), as well as full tree files (as we've generated them).

3. In the files folder, for every tree_id, calculate the minimum between our local generation and on chain generation. If the files miss, we write them, and try again after decrementing the generation, until the files are present in the folder. This guarantees we write the files for the data only owned by us, but not until it's fully confirmed on chain.

4. `data_layer_server.py` is taking as a query the filename and if it exists in the watched foldername, it blindly serves it.

Draft for:
- [x] Just git mechanics around the pair of atari-related branches and release paths.  Hopefully this gets merged later today.

### CODE REVIEW

File: chia/data_layer/download_data.py [hunk 1]
```diff
@@ -1,166 +1,141 @@
...
+        existing_generation += 1
+        filename = get_delta_filename(tree_id, root_hash, existing_generation)
+        url = f"http://{ip}:{port}/{filename}"
 
-    return True
+        try:
+            async with aiohttp.ClientSession() as session:
+                async with session.get(url) as resp:
+                    if resp.status != 200:
+                        raise RuntimeError("Didn't get 200 response status.")
```
Comment (altendky) [inline chia/data_layer/download_data.py:99]: https://docs.aiohttp.org/en/stable/client_reference.html?highlight=raise%20for%20status#aiohttp.ClientResponse.raise_for_status

```suggestion
                    resp.raise_for_status()
```

I think you could alternatively just do `aiohttp.ClientSession(raise_for_status=True)` above.

File: chia/data_layer/download_data.py [hunk 2]
```diff
@@ -1,166 +1,141 @@
...
 
+                    target_filename = os.path.join(client_foldername, filename)
```
Comment (altendky) [inline chia/data_layer/download_data.py:101]:
```suggestion
                    target_filename = client_foldername.joinpath(filename)
```

File: chia/data_layer/download_data.py [hunk 3]
```diff
@@ -1,166 +1,141 @@
...
+                    with open(target_filename, "wb") as writer:
+                        text = await resp.read()
+                        writer.write(text)
```
Comment (altendky) [inline chia/data_layer/download_data.py:104]: There are probably several places to use the `Path` methods.

```suggestion
                    text = await resp.read()
                    target_filename.write_bytes(text)
```

File: chia/data_layer/download_data.py [hunk 4]
```diff
@@ -1,166 +1,141 @@
...
+        except asyncio.CancelledError:
+            raise
+        except Exception:
```
Comment (altendky) [inline chia/data_layer/download_data.py:107]: For any exception?  Even a `NameError` or such?

File: chia/data_layer/util/benchmark.py [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
-from chia.types.blockchain_format.sized_bytes import bytes32
+from typing import Dict, Optional
+
+import aiosqlite
+
 from chia.data_layer.data_layer_types import Side, TerminalNode
+from chia.data_layer.data_store import DataStore
 from chia.types.blockchain_format.program import Program
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.util.db_wrapper import DBWrapper
```
Comment (altendky) [inline chia/data_layer/util/benchmark.py:15]: In general, let's leave application of isort for dedicated PRs.

File: chia/rpc/data_layer_rpc_api.py [hunk 6]
```diff
@@ -198,20 +200,13 @@ async def subscribe(self, request: Dict[str, Any]) -> Dict[str, Any]:
...
-        req_mode = request.get("mode")
-        if req_mode is not None:
-            mode = DownloadMode(req_mode)
+
         if self.service is None:
             raise Exception("Data layer not created")
         store_id_bytes = bytes32.from_hexstr(store_id)
-        await self.service.subscribe(store_id=store_id_bytes, mode=mode, ip=ip, port=port)
+        ips = request["ip"]
+        ports = request["port"]
```
Comment (altendky) [inline chia/rpc/data_layer_rpc_api.py:208]: As above, I don't think that parallel lists are good.

File: tests/core/data_layer/test_data_store.py [hunk 7]
```diff
@@ -325,6 +335,89 @@ async def test_get_ancestors_optimized(data_store: DataStore, tree_id: bytes32)
...
+    "use_optimized",
+    [True, False],
+)
+async def test_batch_update(data_store: DataStore, tree_id: bytes32, use_optimized: bool) -> None:
+    num_batches = 10
+    num_ops_per_batch = 100 if use_optimized else 10
+    saved_roots: List[Root] = []
+    saved_batches: List[List[Dict[str, Any]]] = []
+
+    with tempfile.TemporaryDirectory() as temp_directory:
```
Comment (altendky) [inline tests/core/data_layer/test_data_store.py:349]: Use the `tmp_path` fixture for a temporary directory.

**Review by altendky — CHANGES_REQUESTED** [review]

File: chia/data_layer/data_layer.py [hunk 8]
```diff
@@ -102,33 +108,24 @@ async def batch_update(
...
-                side = change.get("side")
-                if reference_node_hash or side:
-                    await self.data_store.insert(key, value, tree_id, reference_node_hash, side, hint_keys_values)
-                await self.data_store.autoinsert(key, value, tree_id, hint_keys_values)
+        old_root = await self.data_store.get_tree_root(tree_id=tree_id)
+        rollback_generation = 0 if old_root is None else old_root.generation
+        try:
+            await self.data_store.insert_batch(tree_id, changelist)
+            root = await self.data_store.get_tree_root(tree_id=tree_id)
+            # todo return empty node hash from get_tree_root
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py:119]: this is old code (but it appears new due to changing tabs) - I think the point is for `data_store` to return `bytes32([0]*32)` instead of `None`, so we are consistent with wallet. I think Almog did some work here on this:
https://github.com/Chia-Network/chia-blockchain/pull/10475

File: chia/data_layer/data_layer.py [hunk 9]
```diff
@@ -102,33 +108,24 @@ async def batch_update(
...
-        else:
-            node_hash = self.none_bytes  # todo change
-        transaction_record = await self.wallet_rpc.dl_update_root(tree_id, node_hash, fee)
-        assert transaction_record
-        # todo register callback to change status in data store
-        # await self.data_store.change_root_status(root, Status.COMMITTED)
-        return transaction_record
+                node_hash = self.none_bytes  # todo change
+            transaction_record = await self.wallet_rpc.dl_update_root(tree_id, node_hash, fee)
+            assert transaction_record
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py:125]: still old code that appears new - asserting that wallet was able to make the transaction containing the batch update I think.

File: chia/data_layer/data_layer.py [hunk 10]
```diff
@@ -224,105 +183,91 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
...
-        await self.data_store.set_validated_wallet_generation(tree_id, int(singleton_record.generation))
-
-    async def subscribe(self, store_id: bytes32, mode: DownloadMode, ip: str, port: uint16) -> None:
-        subscription = Subscription(store_id, mode, ip, port)
+                self.log.warning(f"Exception while downloading files for {tree_id}: {e} {traceback.format_exc()}.")
+
+    async def upload_files(self, tree_id: bytes32) -> None:
+        singleton_record: Optional[SingletonRecord] = await self.wallet_rpc.dl_latest_singleton(tree_id, True)
+        if singleton_record is None:
+            return
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py:242]: it's possible we've just created the store - so the `singleton_record` isn't confirmed yet on chain when we get to this function. But I agree we should at least log this case.

File: chia/data_layer/data_layer.py [hunk 11]
```diff
@@ -224,105 +183,91 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
...
+        root = await self.data_store.get_tree_root(tree_id=tree_id)
+        publish_generation = min(int(singleton_record.generation), 0 if root is None else root.generation)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py:244]: yes, since `batch_update` makes now 1 generation to be 1 batch, not 1 single operation.

File: chia/data_layer/data_layer.py [hunk 12]
```diff
@@ -359,25 +305,48 @@ async def periodically_fetch_data(self) -> None:
                     break
                 except aiohttp.client_exceptions.ClientConnectorError:
                     pass
+                except asyncio.CancelledError:
+                    return
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py:309]: this is the main task (so nothing awaits it, except for the DataLayer object). I've tried to return in cases like this and raise otherwise (where something else awaits the function).

File: chia/data_layer/data_store.py [hunk 13]
```diff
@@ -843,6 +865,98 @@ async def delete(
...
+    async def insert_batch(
+        self,
+        tree_id: bytes32,
+        changelist: List[Dict[str, Any]],
+        status: Status = Status.PENDING,
+        *,
+        lock: bool = True,
+    ) -> None:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            hint_keys_values = await self.get_keys_values_dict(tree_id, lock=False)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:877]: yes, all terminal nodes will be in memory. this is still old code - I've just moved it from data_layer to here, since it looked more appropriate, so it worked like this before. the point is to optimise looking up a value by key - needed by `delete`. without this we had very slow performance.

File: chia/data_layer/data_store.py [hunk 14]
```diff
@@ -843,6 +865,98 @@ async def delete(
...
+            old_root = await self.get_tree_root(tree_id, lock=False)
+            for change in changelist:
+                if change["action"] == "insert":
+                    key = change["key"]
+                    value = change["value"]
+                    reference_node_hash = change.get("reference_node_hash")
+                    side = change.get("side")
+                    if reference_node_hash or side:
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:885]: no, if we provide either of them we should enter the insert mechanism. although I see that providing only side, without a reference_node will not work properly (I guess the behaviour is to pick a random node, and insert on the left of it?).

File: chia/data_layer/data_store.py [hunk 15]
```diff
@@ -843,6 +865,98 @@ async def delete(
...
+                    else:
+                        await self.autoinsert(key, value, tree_id, hint_keys_values, lock=False)
+                else:
+                    assert change["action"] == "delete"
+                    key = change["key"]
+                    await self.delete(key, tree_id, hint_keys_values, lock=False)
+
+            root = await self.get_tree_root(tree_id, lock=False)
+            # We delete all "temporary" records stored in root and ancestor tables and store only the final result.
+            await self.rollback_to_generation(tree_id, old_root.generation, lock=False)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:896]: yes - but only for the writer. the clients will download only the path from the root. the fix requires some extra work: traverse all intermediate roots, store all nodes, and then delete the nodes that are not on the path from the final root. I'm not sure this is worth it since all those tree crossings and deletions will contribute to the runtime of `batch_update`.

File: chia/data_layer/data_store.py [hunk 16]
```diff
@@ -940,187 +1054,79 @@ async def get_proof_of_inclusion_by_key(
...
 
-        return result
+        with open(filename, "ab") as writer:
+            writer.write(len(to_write).to_bytes(4, byteorder="big"))
+            writer.write(to_write)
 
     async def subscribe(self, subscription: Subscription, *, lock: bool = True) -> None:
         async with self.db_wrapper.locked_transaction(lock=lock):
+            servers_ip = json.dumps(subscription.ip)
+            servers_port = json.dumps(subscription.port)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:1106]: I felt it's more natural given how we update the subscriptions - in the DB case we'd have to delete everything each time and then reinsert the new data.

File: chia/data_layer/download_data.py [hunk 17]
```diff
@@ -1,166 +1,141 @@
...
+            serialized_node = SerializedNode.from_bytes(serialize_nodes_bytes)
+
+            if serialized_node.is_terminal:
+                await data_store.insert_node(NodeType.TERMINAL, serialized_node.value1, serialized_node.value2)
+            else:
+                await data_store.insert_node(NodeType.INTERNAL, serialized_node.value1, serialized_node.value2)
+
+    async with data_store.db_wrapper.locked_transaction(lock=True):
+        await data_store._insert_root(tree_id=tree_id, node_hash=root_hash, status=Status.COMMITTED)
+    await data_store.build_ancestor_table_from_root(tree_id)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/download_data.py:44]: no, `build_ancestor_table_from_root` has its own transaction.

File: chia/data_layer/data_store.py [hunk 18]
```diff
@@ -940,187 +1054,79 @@ async def get_proof_of_inclusion_by_key(
...
-                    continue
-                else:
-                    operation = await self.get_single_operation(
-                        previous_root.node_hash, current_root.node_hash, current_root.status, lock=False
-                    )
-                    result.append(operation)
-                previous_root = current_root
+    ) -> None:
+        if node_hash == bytes32([0] * 32):
+            open(filename, "ab").close()
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:1081]: in this case it is the empty tree (hence the empty file): `bytes32([0] * 32)`

File: chia/data_layer/data_store.py [hunk 19]
```diff
@@ -940,187 +1054,79 @@ async def get_proof_of_inclusion_by_key(
...
+        if isinstance(node, InternalNode):
+            await self.write_tree_to_file(root, node.left_hash, tree_id, deltas_only, filename, lock=lock)
+            await self.write_tree_to_file(root, node.right_hash, tree_id, deltas_only, filename, lock=lock)
+            to_write = bytes(SerializedNode(False, bytes(node.left_hash), bytes(node.right_hash)))
+        else:
+            assert isinstance(node, TerminalNode)
+            to_write = bytes(SerializedNode(True, node.key, node.value))
 
-        return result
+        with open(filename, "ab") as writer:
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:1099]: we want to append to the file even if it doesn't exist - this happens for the first terminal node the depth first search finds.

File: chia/data_layer/data_layer_server.py [hunk 20]
```diff
@@ -1,151 +1,60 @@
...
-            if isinstance(node, TerminalNode):
-                answer.append(
-                    {
-                        "key": node.key.hex(),
-                        "value": node.value.hex(),
-                        "is_terminal": True,
-                    }
+    async def file_handler(self, request: web.Request) -> web.Response:
+        filename = request.match_info["filename"]
+        file_path = os.path.join(self.server_dir, filename)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer_server.py]: I've added `is_filename_valid` method (which checks the input respects our format).

File: chia/data_layer/data_layer.py [hunk 21]
```diff
@@ -102,33 +108,24 @@ async def batch_update(
...
-        # await self.data_store.change_root_status(root, Status.COMMITTED)
-        return transaction_record
+                node_hash = self.none_bytes  # todo change
+            transaction_record = await self.wallet_rpc.dl_update_root(tree_id, node_hash, fee)
+            assert transaction_record
+            # todo register callback to change status in data store
+            # await self.data_store.change_root_status(root, Status.COMMITTED)
+            return transaction_record
+        except Exception:
+            await self.data_store.rollback_to_generation(tree_id, rollback_generation)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py]: I've reused `self.db_wrapper` (and removed my rollback code), but I'm not sure if this is what you meant.

File: chia/data_layer/data_layer.py [hunk 22]
```diff
@@ -102,33 +108,22 @@ async def batch_update(
...
-        for change in changelist:
-            if change["action"] == "insert":
-                key = change["key"]
-                value = change["value"]
-                reference_node_hash = change.get("reference_node_hash")
-                side = change.get("side")
-                if reference_node_hash or side:
-                    await self.data_store.insert(key, value, tree_id, reference_node_hash, side, hint_keys_values)
-                await self.data_store.autoinsert(key, value, tree_id, hint_keys_values)
+        async with self.db_wrapper.locked_transaction(lock=True):
```
Comment (altendky) [inline chia/data_layer/data_layer.py:111]: Is this the data store's db?  I don't think we should be interacting with that outside of the store class itself.

File: chia/data_layer/data_layer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_layer.py]: I thought the wallet generations progressed for non-update spends made by other people to get the singleton announcements.

File: chia/data_layer/data_layer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_layer.py:305]: I think the thing calling this should manage the cancellation, if even that.  Otherwise any time we reuse code, maybe for tests, we have to deal with the fact it doesn't cancel properly.  Code should not cancel termination unless it initiated the cancellation of sub-code.  Asyncio doesn't offer a means by which to differentiate this accurately at this point so at best it is pretty hard to do.

File: chia/data_layer/data_layer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py]: I see - so I suppose the root hash remains the same. in this case we'll have empty delta files generated in this step (and getting the full tree pasted in full tree files). downloading part will simply reinsert the same root hash, but with the incremented generation. if the root hash changes after the non-update operations, this will be problematic.

File: chia/data_layer/data_store.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_store.py:907]: `Side.LEFT` is an instance of a subclass of `int` with a value of 0.  It is falsey.  If you mean to check if something is or is not `None` it is good to reference `None` in the check.

File: chia/data_layer/data_store.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_store.py:1111]: `a` is append.  If the file already exists with data it won't be truncated.  It seems like we should either truncate or raise an exception.

File: chia/data_layer/data_store.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_store.py:1130]: I suspect that we really do not want to reopen a file every time we have a node to write.  Also, as long as we maintain balance in our tree recursion will be ok here, but in general recursion is not a great option in Python.  Function calls are expensive and there's a stack depth limit below 1000.  It's been awhile but can we not have a query with a cursor we walk and write bits as we get them?

I expect that not reopening the file is likely a required change for performance.  Not being recursive a fair bit less important.

File: chia/data_layer/data_store.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_store.py]: We can't just add a subscription or remove one through the RPC?  We have to read them all out, modify as we want, then write them all back?

File: chia/data_layer/download_data.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/download_data.py:63]: But if we fail partway through it won't remove the newly inserted root.  It will just leave the ancestors table incomplete.

File: chia/data_layer/download_data.py [hunk 23]
```diff
@@ -1,166 +1,159 @@
...
+    else:
+        node_hash = bytes32([0] * 32)  # todo change
+    filename_full_tree = foldername.joinpath(get_full_tree_filename(tree_id, node_hash, root.generation))
+    filename_diff_tree = foldername.joinpath(get_delta_filename(tree_id, node_hash, root.generation))
+    written = False
+    if override and filename_full_tree.exists():
+        os.remove(filename_full_tree)
+    if not filename_full_tree.exists():
+        await data_store.write_tree_to_file(root, node_hash, tree_id, False, filename_full_tree)
+        written = True
```
Comment (altendky) [inline chia/data_layer/download_data.py:84]: This still has a race condition where even with an override it can still just not write anything if the file is recreated after we delete it.  I think we can open with `w` if overriding and `x` and handle the _specific_ exception and number.  Maybe double check the truncation behavior with `x`.

Same below.

```pycon
>>> f = open("abcd", "x")
>>> f = open("abcd", "x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileExistsError: [Errno 17] File exists: 'abcd'
```

File: tests/core/data_layer/test_data_store.py [hunk 24]
```diff
@@ -1119,17 +1097,72 @@ async def test_rollback_to_generation(data_store: DataStore, tree_id: bytes32) -
...
+    "test_delta",
+    [True, False],
+)
+@pytest.mark.asyncio
+async def test_data_server_files(data_store: DataStore, tree_id: bytes32, test_delta: bool, tmpdir: Path) -> None:
+    roots: List[Root] = []
+    num_batches = 10
+    num_ops_per_batch = 100
+
+    temp_directory_path = Path(tmpdir)
```
Comment (altendky) [inline tests/core/data_layer/test_data_store.py:1124]: The `tmp_path` fixture is already a `Path`, if I remember correctly.  Likewise in the test above.

File: chia/data_layer/data_layer.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_layer.py]: ok, I've raised it inside this task as well, and caught it only when doing `self.periodically_manage_data_task.cancel()`

File: chia/data_layer/download_data.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/download_data.py]: ok, I've moved inserting the root together with building the ancestor table. This fixes the awkwardness inside `download_data.py` as well where we had to call the private `_insert_root`.

File: chia/data_layer/data_layer.py (hunk 22 above)
Comment (fchirica) [reply chia/data_layer/data_layer.py]: I've created a new one for now, for each batch_update. What happens if we get another write transaction while working on this batch update? (for example downloading the data from subscription). Since now there are 2 DBWrappers (as opposed to using a single lock as before everywhere), would that wait until we finish this batch update or it will throw an exception?

File: chia/data_layer/data_store.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py]: we'll probably need priorities in this case - i.e. first the main server(s), then the mirrors, then the mirrors of the mirrors etc. I feel how to implement this depends a lot of what CW will do - i.e. if it'll manage all of that and just paste our list, or if it'll take them one by one, so it's probably worth discussing the design further in this case.

[the remark this answers is not in the document]
Comment (altendky) [reply chia/data_layer/data_store.py]: Sounds good.  I don't know what is best but I'll mention the idea of associating priority numbers to each subscription instead of just being order based.  Anyways, @emlowe I guess this needs to get added for discussion and tracking.

File: chia/data_layer/data_layer.py [hunk 25]
```diff
@@ -224,105 +183,92 @@ async def fetch_and_validate(self, subscription: Subscription) -> None:
...
+            self.log.error(f"No singleton record found for: {store_id}")
+            return
+        max_generation = min(singleton_record.generation, 0 if root is None else root.generation)
+        server_files_location = foldername if foldername is not None else self.server_files_location
+        for generation in range(1, max_generation + 1):
+            root = await self.data_store.get_tree_root(tree_id=store_id, generation=generation)
+            await write_files_for_root(self.data_store, store_id, root, server_files_location, override)
+
+    async def subscribe(self, store_id: bytes32, urls: List[str]) -> None:
+        parsed_urls = [url[:-1] if url.endswith("/") else url for url in urls]
```
Comment (altendky) [inline chia/data_layer/data_layer.py:270]: Will remove any trailing `/`s.

```suggestion
        parsed_urls = [url.rstrip("/") for url in urls]
```

File: .isort.cfg [hunk 26]
```diff
@@ -38,6 +38,7 @@ extend_skip=
     chia/data_layer/data_store.py
     chia/data_layer/dl_wallet_store.py
     chia/data_layer/download_data.py
+    chia/data_layer/util/benchmark.py
```
Comment (altendky) [inline .isort.cfg:41]: I apologize for not realizing this was a rename when I complained about the isort'ing.  So two choices from here now that I get that.  Regardless we should delete the old name from this file since it is no longer needed.  From there you can remove this line too and do the isort now that I get that it is localized and there's a reason for it, or just leave this be.  Maybe next time the whole move can be a separate PR since it seem unrelated.  This is fine for now though.

File: chia/data_layer/download_data.py [hunk 27]
```diff
@@ -1,166 +1,189 @@
...
+    mode = "wb" if override else "xb"
+
+    try:
+        with open(filename_full_tree, mode) as writer:
+            await data_store.write_tree_to_file(root, node_hash, tree_id, False, writer)
+        written = True
+    except FileExistsError:
+        pass
+    except Exception:
+        raise
```
Comment (altendky) [inline chia/data_layer/download_data.py:117]:
```suggestion
```

File: chia/data_layer/data_store.py [hunk 28]
```diff
@@ -950,187 +1075,68 @@ async def get_proof_of_inclusion_by_key(
...
+    async def write_tree_to_file(
         self,
-        batch: List[Tuple[NodeType, str, str]],
+        root: Root,
+        node_hash: bytes32,
         tree_id: bytes32,
-        root_hash: bytes32,
-        generation: int,
+        deltas_only: bool,
+        writer: IO[Any],
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/data_store.py:1097]: it previously didn't like opening with variable `mode` - but if I make the option explicit instead of using variable `mode` it seems fine now.

File: chia/data_layer/download_data.py [hunk 29]
```diff
@@ -1,166 +1,189 @@
...
+    filename: Path,
+) -> None:
+    with open(filename, "rb") as reader:
+        while True:
+            chunk = b""
+            while len(chunk) < 4:
+                size_to_read = 4 - len(chunk)
+                cur_chunk = reader.read(size_to_read)
+                if cur_chunk is None or cur_chunk == b"":
+                    break
```
[the remark this answers is not in the document]
Comment (fchirica) [reply chia/data_layer/download_data.py:62]: yes, we need to raise every time except when we have a full chunk we couldn't read - this means EOF.

**Review by altendky — APPROVED** [review]
Approving, but going to mark draft for the moment as we aren't going to merge it quite yet.  Hopefully later today.  See discussion on keybase for an explanation of the git mechanics working towards a 1.4.0+atari release once 1.4.0 is out but without this new server.
